include/util/for_each_range.hpp

```diff
@@ -0,0 +1,27 @@
+#ifndef OSRM_UTIL_FOR_EACH_RANGE_HPP
+#define OSRM_UTIL_FOR_EACH_RANGE_HPP
+
+namespace osrm
+{
+namespace util
+{
+
+template <typename Iter, typename Func> void for_each_range(Iter begin, Iter end, Func f)
+{
+    auto iter = begin;
+    while (iter != end)
+    {
+        const auto key = iter->first;
+        auto begin_range = iter;
+        while (iter != end && iter->first == key)
```

Comment:
Not sure if this should be a generic util function then at all. Even now it depends on the sub ranges being represented in pairs.